Description: Port functionality of comp script needed by opaconfig
 opaconfig allows a package to override certain funcionality. This
 patch ports the required functionality of .comp_ofed_delta to Debian.
 Omit config directory deletion in uninstall_opa_stack, let dpkg manage it.
Author: Brian T. Smith <bsmith@systemfabricworks.com>
Forwarded: not-needed
Last-Update: <2018-02-21>
---
This patch header follows DEP-3: http://dep.debian.net/deps/dep3/
--- a/.comp_ofed_delta.pl
+++ b/.comp_ofed_delta.pl
@@ -246,6 +246,23 @@
 				"delta_debug",		# must be last real component
 );
 
+my @delta_components_debian8 = (
+				"opa_stack", 		# Kernel drivers.
+				"ibacm", 			# OFA IB communication manager assistant.
+				"intel_hfi", 		# HFI drivers
+				"delta_ipoib", 		# ipoib module.
+				"mpi_selector",
+				"mvapich2",
+				"openmpi",
+				"gasnet",
+				"openshmem",
+				"sandiashmem",
+				"opa_stack_dev", 	# dev libraries.
+				"delta_mpisrc", 	# Source bundle for MPIs.
+				"hfi1_uefi",
+				"delta_debug",		# must be last real component
+	);
+
 my @delta_components_rhel72 = ( @delta_components_other );
 
 my @delta_components_rhel67 = ( @delta_components_other );
@@ -596,6 +613,18 @@
 					},
 );
 
+my %opa_stack_dev_debian8_comp_info = (
+	'opa_stack_dev' => {
+					KernelRpms => [ ],
+					UserRpms =>  [ "ifs-kernel-updates-devel", "ibacm-devel",
+						"libibumad-devel", "libibmad-devel"],
+					DebugRpms =>  [  ],
+					Drivers => "", 	# none
+					StartupScript => "",
+					StartupParams => [ ],
+					},
+);
+
 my %opa_stack_dev_rhel70_comp_info = (
         'opa_stack_dev' => {
                                         KernelRpms => [ "" ],
@@ -729,6 +758,22 @@
 					},
 );
 
+my %opa_stack_debian8_comp_info = (
+	'opa_stack' => {
+					KernelRpms => [ "kmod-ifs-kernel-updates" ], # special case
+					UserRpms =>	  [ "opa-scripts",
+								"srptools",
+								"libibmad", "libibumad", "librdmacm",
+								"infiniband-diags",
+								  ],
+					DebugRpms =>  [ "srptools-debuginfo",
+								  ],
+					Drivers => "",
+					StartupScript => "opa",
+					StartupParams => [ "ARPTABLE_TUNING" ],
+					},
+);
+
 my %opa_stack_rhel67_comp_info = (
 	'opa_stack' => {
 					KernelRpms => [ "ifs-kernel-updates" ], # special case
@@ -863,6 +908,24 @@
 	%delta_debug_comp_info,
 );
 
+my %delta_comp_info_debian8 = (
+	%opa_stack_debian8_comp_info,
+	%ibacm_comp_info,
+	%intel_hfi_comp_info,
+	%ib_wfr_lite_comp_info,
+	%delta_ipoib_comp_info,
+	%mpi_selector_comp_info,
+	%mvapich2_comp_info,
+	%openmpi_comp_info,
+	%gasnet_comp_info,
+	%openshmem_comp_info,
+	%sandiashmem_comp_info,
+	%opa_stack_dev_debian8_comp_info,
+	%delta_mpisrc_comp_info,
+	%hfi1_uefi_comp_info,
+	%delta_debug_comp_info,
+);
+
 my %delta_comp_info_rhel67 = (
 	%opa_stack_rhel67_comp_info,
 	%ibacm_comp_info,
@@ -1027,6 +1090,7 @@
 # all kernel srpms
 # these are in the order we must build/process them to meet basic dependencies
 my @delta_kernel_srpms_other = ( 'compat-rdma' );
+my @delta_kernel_srpms_debian8 = ( 'kmod-ifs-kernel-updates' );
 my @delta_kernel_srpms_rhel72 = ( 'kmod-ifs-kernel-updates' );
 my @delta_kernel_srpms_rhel70 = ( 'compat-rdma' );
 my @delta_kernel_srpms_rhel67 = ( 'ifs-kernel-updates' );
@@ -1045,6 +1109,12 @@
  		"mvapich2", "openmpi", "gasnet", "openshmem", "openshmem-test-suite",
 		"shmem-benchmarks", "srptools", "libibmad", "infiniband-diags", "hfi1_uefi", "libfabric", "sandiashmem"
 );
+my @delta_user_srpms_debian8 = (
+		"opa-scripts", "mpi-selector", "ibacm",
+		"libhfi1", "libpsm2", "hfi1-diagtools-sw", "hfidiags", "hfi1-firmware", "hfi1-firmware_debug",
+ 		"mvapich2", "openmpi","gasnet", "openshmem", "openshmem-test-suite",
+	    "shmem-benchmarks", "srptools", "libibumad", "libibmad", "infiniband-diags", "hfi1_uefi", "libfabric", "sandiashmem"
+);
 my @delta_user_srpms_rhel67 = (
 		"opa-scripts", "libibumad", "ibacm", "mpi-selector",
 		"libhfi1", "libpsm2", "hfi1-diagtools-sw", "hfidiags", "hfi1-firmware", "hfi1-firmware_debug",
@@ -1361,7 +1431,13 @@
 	%delta_rpm_info = ();	# start fresh
 
 	# filter components by distro
-	if ("$CUR_DISTRO_VENDOR" eq 'SuSE'
+	 if ( "$CUR_DISTRO_VENDOR" eq 'debian'
+		  && "$CUR_VENDOR_VER" eq 'ES8' ) {
+		@delta_components = ( @delta_components_debian8 );
+		%delta_comp_info = ( %delta_comp_info_debian8 );
+		@delta_kernel_srpms = ( @delta_kernel_srpms_debian8 );
+		@delta_user_srpms = ( @delta_user_srpms_debian8 );
+	} elsif ("$CUR_DISTRO_VENDOR" eq 'SuSE'
 		&& ("$CUR_VENDOR_VER" eq 'ES12' || "$CUR_VENDOR_VER" eq 'ES121')) {
 		@delta_components = ( @delta_components_sles );
 		%delta_comp_info = ( %delta_comp_info_sles );
@@ -1710,6 +1786,56 @@
 	}
 }
 
+# uninstall debs which are in package_list and are not needed by
+# any components in install_list
+# all variations of the specified packages are uninstalled
+sub delta_deb_uninstall_not_needed_list($$$$@)
+{
+	my $install_list = shift();	# components which will remain on system
+	my $uninstalling_list = shift();	# components which are being uninstalled
+	my $comp = shift();	# component being uninstalled
+	my $verbosity = shift();
+	my(@package_list) = @_;	# package names to consider for uninstall
+
+DEB: foreach my $package ( reverse(@package_list) ) {
+		my @install_list = split /[[:space:]]+/, $install_list;
+		foreach my $c ( @install_list ) {
+			next if ( "$c" eq '' ); # handling leading spaces
+			# see if package is part of a component we are interested in
+			if ( " @delta_components " =~ / $c /
+				 && " $delta_rpm_info{$package}{'PartOf'} " =~ / $c / ) {
+				next DEB;	# its still needed, leave it installed
+			}
+		}
+		if ( $delta_rpm_info{$package}{'Available'} == 0 ) {
+			next DEB; # package was not installed.
+		}
+		# if we get here, package is not in any component we are interested in
+		if ( "$uninstalling_list" ne "" && "$comp" ne "" ) {
+			# we are doing an explicit uninstall, we must be careful
+			# about debs which are part of more than 1 component
+			# uninstalling_list is in dependency order and is executed
+			# backwards, so once we get to processing the 1st component
+			# in uninstalling list which has this package, we know its
+			# safe to remove the package
+			my @uninstalling = split /[[:space:]]+/, $uninstalling_list;
+			foreach my $c ( @uninstalling ) {
+				next if ( "$c" eq '' ); # handling leading spaces
+				if ( " @delta_components " =~ / $c /
+					&& " $delta_rpm_info{$package}{'PartOf'} " =~ / $c / ) {
+					# found 1st uninstalled component with package
+					if ("$c" ne "$comp") {
+						next DEB;	# don't remove til get to $c's uninstall
+					} else {
+						last;	# exit this loop and uninstall package
+					}
+				}
+			}
+		}
+		dpkg_uninstall($package, "any", " --nodeps ", $verbosity);
+	}
+}
+
 
 # resolve filename within $srcdir/$SRPMS_SUBDIR
 # and return filename relative to $srcdir
@@ -2769,6 +2895,24 @@
 				 	$verbosity, @{ $delta_comp_info{$comp}{'KernelRpms'}});
 }
 
+# uninstall all debs associated with an OFED Delta component
+sub uninstall_delta_comp_debs($$$$)
+{
+	my $comp = shift();
+	my $install_list = shift();
+	my $uninstalling_list = shift();
+	my $verbosity = shift();
+
+	# debuginfo never in >1 component, so do explicit uninstall since
+	# have an odd PartOf relationship which confuses uninstall_not_needed_list
+	dpkg_uninstall_list2("any", " --force-depends ",
+						 @{ $delta_comp_info{$comp}{'DebugRpms'}});
+	delta_deb_uninstall_not_needed_list($install_list, $uninstalling_list, $comp,
+				 	$verbosity, @{ $delta_comp_info{$comp}{'UserRpms'}});
+	delta_deb_uninstall_not_needed_list($install_list, $uninstalling_list, $comp,
+				 	$verbosity, @{ $delta_comp_info{$comp}{'KernelRpms'}});
+}
+
 # helper which does most of the work for uninstalling rpms and drivers
 # for an OFED component
 # caller must handle any non-RPM files
@@ -2779,7 +2923,11 @@
 	my $uninstalling_list = shift();
 	my $verbosity = shift();
 
-	uninstall_delta_comp_rpms($comp, $install_list, $uninstalling_list, $verbosity);
+	if ("$CUR_DISTRO_VENDOR" eq 'debian') {
+		uninstall_delta_comp_debs($comp, $install_list, $uninstalling_list, $verbosity);
+	} else {
+		uninstall_delta_comp_rpms($comp, $install_list, $uninstalling_list, $verbosity);
+	}
 	remove_delta_kernel_ib_drivers($comp, $verbosity);
 }
 
@@ -2882,7 +3030,10 @@
 sub installed_delta_opa_stack()
 {
 	my $driver_subdir=$ComponentInfo{'opa_stack'}{'DriverSubdir'};
-	if ( "$CUR_VENDOR_VER" eq "ES67" ) {
+	if ( "$CUR_DISTRO_VENDOR" eq "debian" ) {
+		return ( -e "$ROOT$BASE_DIR/version_delta"
+				&& dpkg_is_installed("kmod-ifs-kernel-updates", $CUR_OS_VER));
+	} elsif ( "$CUR_VENDOR_VER" eq "ES67" ) {
 		return ( -e "$ROOT$BASE_DIR/version_delta" 
 				&& rpm_is_installed("libibumad", "user")
 				&& rpm_is_installed("ifs-kernel-updates", $CUR_OS_VER));
@@ -3109,18 +3260,9 @@
 	#remove_blacklist("ib_ipath");
 
 	uninstall_delta_comp('opa_stack', $install_list, $uninstalling_list, 'verbose');
-	remove_driver_dirs($driver_subdir);
-	#remove_modules_conf;
-	remove_limits_conf;
-
+	dpkg_uninstall('qperf', "any", " --nodeps ", 'verbose');
 	remove_udev_permissions;
 
-	system("rm -rf $ROOT$BASE_DIR/version_delta");
-	system("rm -rf $ROOT/usr/lib/opa/.comp_delta.pl");
-	system "rmdir $ROOT/usr/lib/opa 2>/dev/null";	# remove only if empty
-	system "rmdir $ROOT$BASE_DIR 2>/dev/null";	# remove only if empty
-	system "rmdir $ROOT$OPA_CONFIG_DIR 2>/dev/null";	# remove only if empty
-
 	need_reboot();
 	$ComponentWasInstalled{'opa_stack'}=0;
 }
@@ -3163,7 +3305,11 @@
 sub installed_intel_hfi()
 {
     my $driver_subdir=$ComponentInfo{'intel_hfi'}{'DriverSubdir'};
-    if ( "$CUR_VENDOR_VER" eq "ES67" ) {
+	if ( "$CUR_DISTRO_VENDOR" eq "debian" ) {
+		return ( -e "$ROOT$BASE_DIR/version_delta"
+				&& dpkg_is_installed("libhfi1", "user")
+				&& dpkg_is_installed("kmod-ifs-kernel-updates", $CUR_OS_VER));
+    } elsif ( "$CUR_VENDOR_VER" eq "ES67" ) {
 	return ( -e "$ROOT$BASE_DIR/version_delta"
 			&& rpm_is_installed("libhfi1", "user")
                         && rpm_is_installed("ifs-kernel-updates", $CUR_OS_VER));
@@ -3823,7 +3969,7 @@
 
 sub installed_openmpi()
 {
-	return ((rpm_is_installed("openmpi_gcc_hfi", "user")
+	return ((rpm_is_installed("openmpi_gcc", "user")
 			&& -e "$ROOT$BASE_DIR/version_delta"));
 }
 
@@ -4219,11 +4365,20 @@
 # check file_glob("$srcdir/SRPMS/mvapich-*.src.rpm") ne ""
 # check file_glob("$srcdir/SRPMS/mvapich2-*.src.rpm") ne ""
 # check file_glob("$srcdir/SRPMS/openmpi-*.src.rpm") ne ""
+	if ("$CUR_DISTRO_VENDOR" eq 'debian') {
+		return ( (-d "$srcdir" ) );
+	}
 	return ( (-d "$srcdir/SRPMS" || -d "$srcdir/RPMS" ) );
 }
 
 sub installed_delta_mpisrc()
 {
+	if ("$CUR_DISTRO_VENDOR" eq 'debian') {
+		return ((-e "$ROOT$BASE_DIR/version_delta"
+				 && file_glob("$ROOT/usr/src/opa/MPI/mvapich*.dsc") ne ""
+				 && file_glob("$ROOT/usr/src/opa/MPI/openmpi*.dsc") ne ""
+				 && file_glob("$ROOT/usr/src/opa/MPI/mpitests*.dsc") ne ""));
+	}
 	return ((-e "$ROOT$BASE_DIR/version_delta"
 			&& file_glob("$ROOT/usr/src/opa/MPI/mvapich*.src.rpm") ne ""
 			&& file_glob("$ROOT/usr/src/opa/MPI/openmpi*.src.rpm") ne ""
@@ -4360,6 +4515,10 @@
 
 sub installed_delta_debug()
 {
+	if ("$CUR_DISTRO_VENDOR" eq 'debian') {
+		return (dpkg_is_installed("libibumad3-dbg", "user")
+				&& -e "$ROOT$BASE_DIR/version_delta");
+	}
 	return (rpm_is_installed("libibumad-debuginfo", "user")
 			&& -e "$ROOT$BASE_DIR/version_delta");
 }
